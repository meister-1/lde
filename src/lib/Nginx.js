import fs from "fs-sync"

export default class Nginx {

    constructor(apps, outPath) {
        this.apps = apps.map(el => {
           const ports = el.port.split(":")
           return {
               ...el,
               port: ports.length > 1 ? ports[1] : el.port
            }
        })
        this.outPath = outPath
        this.config = Nginx.getDefaultConfig()
    }

    static getDefaultConfig() {
        const defaultServerConfig = `server {
            listen       80;
            
            location / {
                return 301 https://$host$request_uri;
            }
        }`

        return defaultServerConfig
    } 

    createConfig() {
        
        const nginxServerConfigs = this.apps.map(({domain, port, ip, certs}) => {

            const listenConfig = certs ? `listen      443 ssl;
            ssl_certificate     /etc/nginx/${certs.crt};
            ssl_certificate_key /etc/nginx/${certs.key};` : "listen      80"

                return `server {
            ${listenConfig}
            server_name  ${domain};
        
            location / {
                proxy_pass "http://${ip}:${port}";
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";
            }
        }`
        
        })
        
        this.config = `${Nginx.getDefaultConfig()}\n${nginxServerConfigs.join("\n")}\n`

        return this
        
    }

    getConfig() {
        return this.config
    }

    writeFile() {
        if(this.config) {
            fs.write(`${this.outPath}/nginx/default.conf`, this.getConfig(), function(err) {
                if(err) {
                    console.log(err)
                }
            })
        }
    }
}