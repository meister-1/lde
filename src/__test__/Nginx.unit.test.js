import Nginx from "../lib/Nginx"

const mockOneApp = [{
    ip: "10.5.0.10",
    domain: "a.meister1.com",
    port: "8000",
    sourceDir: "out/a/src",
    name: "a",
    version: "latest",
    gitUrl: "git@bitbucket.org:lokalleads/a.git",
    volumes: ["maven-data:/.m2"]
}]

const mockMultiApp = [{
  ip: "10.5.0.10",
  domain: "a.meister1.com",
  port: "8000",
  sourceDir: "out/a/src",
  name: "a",
  certs: {
    crt: "wildcard-meister1.com.pem",
    key: "wildcard-meister1-key.com.pem"
  },
  version: "latest",
  gitUrl: "git@bitbucket.org:lokalleads/a.git",
  volumes: ["maven-data:/.m2"]
},
{
  ip: "10.5.0.11",
  domain: "b.meister1.com",
  port: "8001",
  sourceDir: "out/b/src",
  name: "b",
  version: "latest",
  gitUrl: "git@bitbucket.org:lokalleads/b.git",
  volumes: ["maven-data:/.m2"],
  variables: `PORT=8001\nFOO=ABC`
}]


describe('nginx test', () => {
    test('create server config with one app', () => {
      expect(new Nginx(mockOneApp).createConfig().getConfig()).toMatchSnapshot()
    })
    test('create server config with multi apps', () => {
      expect(new Nginx(mockMultiApp).createConfig().getConfig()).toMatchSnapshot()
    })
})