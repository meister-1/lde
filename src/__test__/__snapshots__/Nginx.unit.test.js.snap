// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nginx test create server config with multi apps 1`] = `
"server {
            listen       80;
            
            location / {
                return 301 https://$host$request_uri;
            }
        }
server {
            listen      443 ssl;
            ssl_certificate     /etc/nginx/wildcard-meister1.com.pem;
            ssl_certificate_key /etc/nginx/wildcard-meister1-key.com.pem;
            server_name  a.meister1.com;
        
            location / {
                proxy_pass \\"http://10.5.0.10:8000\\";
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection \\"upgrade\\";
            }
        }
server {
            listen      80
            server_name  b.meister1.com;
        
            location / {
                proxy_pass \\"http://10.5.0.11:8001\\";
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection \\"upgrade\\";
            }
        }
"
`;

exports[`nginx test create server config with one app 1`] = `
"server {
            listen       80;
            
            location / {
                return 301 https://$host$request_uri;
            }
        }
server {
            listen      80
            server_name  a.meister1.com;
        
            location / {
                proxy_pass \\"http://10.5.0.10:8000\\";
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection \\"upgrade\\";
            }
        }
"
`;
